clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java

```diff
@@ -0,0 +1,347 @@
+ * say "An un-windowed cumulative total maintained over all time.". The standalone Total metrics in
+ * the codebase seem to be cumulative metrics that will always increase. The Total metric underlying
+ * Meter type is mostly a Total of a Count metric.
+ * We can assume that a Total metric always increases (but it is not guaranteed as the sample values might be both
+ * negative or positive).
+ * For now, Total is converted to CUMULATIVE_DOUBLE unless we find a valid counter-example.
+ *
+ * <p>
+ *
+ * The Sum as it is a sample sum which is not a cumulative metric. It is converted to GAUGE_DOUBLE.
+ *
+ * <p>
+ *
+ * The compound metrics are virtual metrics. They are composed of simple types or anonymous measurable types
+ * which are reported. A compound metric is never reported as-is.
+ *
+ * <p>
+ *
+ * A Meter metric is always created with and reported as 2 KafkaExporter metrics: a rate and a
+ * count. For eg: org.apache.kafka.common.network.Selector has Meter metric for "connection-close" but it
+ * has to be created with a "connection-close-rate" metric of type rate and a "connection-close-total"
+ * metric of type total. So, we will never get a KafkaExporter metric with type Meter.
+ *
+ * <p>
+ *
+ * Frequencies is created with a array of Frequency objects. When a Frequencies metric is registered, each
+ * member Frequency object is converted into an anonymous Measurable and registered. So, a Frequencies metric
+ * is reported with a set of measurables with name = Frequency.name(). As there is no way to figure out the
+ * compound type, each component measurables is converted to a GAUGE_DOUBLE.
+ *
+ * <p>
+ *
+ * Percentiles work the same way as Frequencies. The only difference is that it is composed of Percentile
+ * types instead. So, we should treat the component measurable as GAUGE_DOUBLE.
+ *
+ * <p>
+ *
+ * Some metrics are defined as either anonymous inner classes or lambdas implementing the Measurable
+ * interface. As we do not have any information on how to treat them, we should fallback to treating
+ * them as GAUGE_DOUBLE.
+ *
+ * <p>
+ *
+ * KafkaExporter -> OpenTelemetry mapping for measurables
+ * Avg / Rate / Min / Max / Total / Sum -> Gauge
+ * Count -> Sum
+ * Meter has 2 elements :
+ *  Total -> Sum
+ *  Rate -> Gauge
+ * Frequencies -> each component is Gauge
+ * Percentiles -> each component is Gauge
+ */
+public class KafkaMetricsCollector implements MetricsCollector {
+
+    private static final Logger log = LoggerFactory.getLogger(KafkaMetricsCollector.class);
+
+    private final StateLedger ledger;
+    private final Clock clock;
+    private final MetricNamingStrategy<MetricName> metricNamingStrategy;
+
+    private static final Field METRIC_VALUE_PROVIDER_FIELD;
+
+    static {
+        try {
+            METRIC_VALUE_PROVIDER_FIELD = KafkaMetric.class.getDeclaredField("metricValueProvider");
+            METRIC_VALUE_PROVIDER_FIELD.setAccessible(true);
```

Comment:
Not a fan on generic (usually indicate that the structure of the code is not well setup) -- why do we need to do this? Can't we find a better way?

I don't see a strict reason why `KafkaMetric` must be `final` so why no relax this, make `metricValueProvider` protected, and add a new class `KafkaMetricsInternal` and add a method `isMeasurable(KafkaMetric metric)` ?

Reply:
That's a valid suggestion but I am not sure if I should do that change right away. @xvrl what do you think, I am sure we would have thought this before.

Reply:
@mjsax The fact that one has to go through those hoops to expose metrics seems to indicate that the KafkaMetrics APIs are somewhat broken. The internal thing is fine, but we should file a ticket to fix it and make it a proper public API so any MetricsReporter plugin can access those fields.

Reply:
we can follow something similar to what the OpenTelemetry instrumentation has done here https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/kafka/kafka-clients/kafka-clients-common/library/src/main/java/io/opentelemetry/instrumentation/kafka/internal/KafkaMetricRegistry.java#L92C46-L97 

to avoid the cost of invoking this on every collection, we can wrap the KafkaMetrics object with something that we can call isMeasurable() on

Reply:
@mjsax Shall I create a jira to track this as seems it will require improving KafkaMetrics APIs?

Reply:
SGTM. For this PR, I would still prefer not to use reflections. -- Doing a follow up KIP SGMT (yes, let's file a AK Jira for tracking).

Reply:
@mjsax I have created this ticket: https://issues.apache.org/jira/browse/KAFKA-15850. I ll add a KIP post 3.7 cutoff date to fix the problem. Resolving the comment for now.